pkg/acl/acl.go

```diff
@@ -0,0 +1,408 @@
+)
+
+const (
+	PermRead    Perm = C.ACL_READ
+	PermWrite        = C.ACL_WRITE
+	PermExecute      = C.ACL_EXECUTE
+)
+
+type ACL struct {
+	lib *libHandle
+	a   C.acl_t
+}
+
+// Entry is an entry in an ACL.
+type Entry struct {
+	a *ACL
+	e C.acl_entry_t
+}
+
+// Permset is a set of permissions.
+type Permset struct {
+	a *ACL
+	p C.acl_permset_t
+}
+
+// Perm represents a permission in the e_perm ACL field
+type Perm int
+
+// Tag represents an ACL e_tag entry
+type Tag int
+
+var ErrSoNotFound = errors.New("unable to open a handle to libacl")
+
+type libHandle struct {
+	handle  unsafe.Pointer
+	libname string
+}
+
+func getHandle() (*libHandle, error) {
+	for _, name := range []string{
+		"libacl.so.1",
+		"libacl.so",
+	} {
+		libname := C.CString(name)
+		defer C.free(unsafe.Pointer(libname))
+		handle := C.dlopen(libname, C.RTLD_LAZY)
+		if handle != nil {
+			h := &libHandle{
+				handle:  handle,
+				libname: name,
+			}
+			return h, nil
+		}
+	}
+	return nil, ErrSoNotFound
+}
+
+func getSymbolPointer(handle unsafe.Pointer, symbol string) (unsafe.Pointer, error) {
+	sym := C.CString(symbol)
+	defer C.free(unsafe.Pointer(sym))
+
+	p := C.dlsym(handle, sym)
+	if p == nil {
+		return nil, fmt.Errorf("error resolving symbol %q")
+	}
+
+	return p, nil
+}
+
+func (h *libHandle) close() error {
+	if r := C.dlclose(h.handle); r != 0 {
+		return fmt.Errorf("error closing %v: %d", h.libname, r)
+	}
+	return nil
+}
+
+// InitACL dlopens libacl and returns an ACL object if successful.
+func InitACL() (*ACL, error) {
+	h, err := getHandle()
+	if err != nil {
+		return nil, err
+	}
+
+	return &ACL{lib: h}, nil
+}
+
+// ParseACL parses a string representation of an ACL.
+func (a *ACL) ParseACL(acl string) error {
+	acl_from_text, err := getSymbolPointer(a.lib.handle, "acl_from_text")
+	if err != nil {
+		return err
+	}
+	cacl := C.CString(acl)
+	defer C.free(unsafe.Pointer(cacl))
+
+	retACL, errno := C.my_acl_from_text(acl_from_text, cacl)
+	if retACL == nil {
+		return fmt.Errorf("error calling acl_from_text: %v", errno)
+	}
+
+	a.a = retACL
+
+	return nil
+}
+
+// Free frees libacl's internal structures and closes libacl.
+func (a *ACL) Free() error {
+	acl_free, err := getSymbolPointer(a.lib.handle, "acl_free")
+	if err != nil {
+		return err
+	}
+
+	ret, errno := C.my_acl_free(acl_free, a.a)
+	if ret < 0 {
+		return fmt.Errorf("error calling acl_free: %v", errno)
+	}
+
+	return a.lib.close()
+}
+
+// SetFileACLDefault sets the "default" ACL for path.
+func (a *ACL) SetFileACLDefault(path string) error {
+	acl_set_file, err := getSymbolPointer(a.lib.handle, "acl_set_file")
+	if err != nil {
+		return err
+	}
+
+	cpath := C.CString(path)
+	defer C.free(unsafe.Pointer(cpath))
+
+	ret, errno := C.my_acl_set_file(acl_set_file, cpath, C.ACL_TYPE_DEFAULT, a.a)
+	if ret < 0 {
+		return fmt.Errorf("error calling acl_set_file: %v", errno)
+	}
+
+	return nil
+}
+
+// Valid checks whether the ACL is valid.
+func (a *ACL) Valid() (bool, error) {
```

Comment:
I wanted to distinguish between an error getting the function or whatever and a real invalid acl